Updated Storage reference page (#7399)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/storage/index.md

- Stored settings persist even when using [split incognito][incognito].
- Includes an exclusive read-only [managed storage area][manifest-storage] for enterprise policies.

## Manifest {: #manifest}

To use the storage API, declare the `"storage"` permission in the extension
[manifest][doc-manifest]. For example:

```json
{
  "name": "My extension",
  ...
  "permissions": [
    "storage"
  ],
  ...
}
```

## Concepts and usage {: #usage }

{% Details %}
{% DetailsSummary %}
Can extensions use web storage APIs?
{% endDetailsSummary %}

While extensions can use the [`Storage`][mdn-storage] interface (accessible from `window.localStorage`) in some contexts (popup and other HTML pages), we don't recommend it for the following reasons:

> - Extension's service worker can't access `Storage`.

Reviewer: ```suggestion
- Extension service workers can't use the Web Storage API.
```

- Content scripts share storage with the host page.

> - Data saved using the `Storage` interface is lost when the user clears their browsing history.

Reviewer: ```suggestion
- Data saved using the Web Storage API is lost when the user clears their browsing history.
```

To move data from web storage APIs to extension storage APIs from a service worker:

1. Create an offscreen document with a conversion routine and an [`onMessage`][on-message] handler.

> 2. Add a conversion routine to an offscreen document. <!--Is 2 an alternative to 1? If so, I'd like to combine them and present them as options.-->
> 3. In the extension service worker, check `chrome.storage` for your data.
> 4. If your data isn't found, [create][create-offscreen] an offscreen document and call [`sendMessage()`][send-message] to start the conversion routine.
> 5. Inside the offscreen document's `onMessage` handler, call the conversion routine.

Reviewer: ```suggestion
1. In the extension service worker, check `chrome.storage` for your data.
1. If your data isn't found, [create][create-offscreen] an offscreen document and call [`sendMessage()`][send-message] to start the conversion routine.
1. Inside the offscreen document's `onMessage` handler, call the conversion routine.
```

Reviewer: I rewrote the first step a while back and neglected to remove the old version.

FYI, the markdown converter numbers automatically when you use '1' for every step. That's what we always do so that we don't need to renumber everything when we change a procedure.

There are also some nuances to how web storage APIs work in extensions. Learn more in the
[Storage and Cookies][storage-and-cookies] article.

{% endDetails %}

### Storage areas

> The Storage API is divided into the following *storage areas*:

Reviewer: ```suggestion
The Storage API is divided into the following storage areas:
```

[`storage.local`][prop-local]

> : Data is stored locally and cleared when the extension is removed. The storage limit is 10 MB (5&nbsp;MB in Chrome 113 and earlier), but can be increased by requesting the `"unlimitedStorage"` permission. We recommend using `storage.local` using it to store larger amounts of data.

Reviewer: ```suggestion
: Data is stored locally and cleared when the extension is removed. The storage limit is 10 MB (5&nbsp;MB in Chrome 113 and earlier), but can be increased by requesting the `"unlimitedStorage"` permission. We recommend using `storage.local` to store larger amounts of data.
```

> <--I think we removed unlimitedStorage from the permissions list. How does that affect this?-->

Reviewer: ```suggestion
```

Reviewer: I messed up when I have you remove `unlimitedStorage`. I just checked [where it's specified](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/api/_permission_features.json;l=654). Can you please add it back.

Author: Done.

[`storage.sync`][prop-sync]
: If syncing is enabled, the data is synced to any Chrome browser that the user is logged into. If disabled, it behaves like `storage.local`. Chrome stores the data locally when the browser is offline and resumes syncing when it's back online. The quota limitation is approximately 100 KB, 8 KB per item. We recommend using `storage.sync` to preserve user settings across synced browsers. If you're working with sensitive user data, instead use `storage.session`.

[`storage.session`][prop-session]

> : Holds data in memory for the duration of a browser session. By default, it's not exposed to content scripts, but this behavior can be changed by setting [`chrome.storage.session.setAccessLevel()`][method-access-level]. The storage limit is 10 MB (1&nbsp;MB in Chrome 111 and earlier). We recommend using `storage.session` to store global variables across service worker runs.

Reviewer: ```suggestion
: Holds data in memory for the duration of a browser session. By default, it's not exposed to content scripts, but this behavior can be changed by setting [`chrome.storage.session.setAccessLevel()`][method-access-level]. The storage limit is 10 MB (1&nbsp;MB in Chrome 111 and earlier). The`storage.session` interface is one of several [we recommend for service workers](/docs/extensions/mv3/service_workers/service-worker-lifecycle/#persist-data).
```

To learn more about the `managed` storage area, see [Manifest for storage areas][manifest-storage].

### Storage and throttling limits {: #storage-and-throttling-limits}

Don't think of adding to the Storage API as putting things in a big truck. Think of adding to
storage as being like putting something in a pipe. The pipe may have material in it already, and it
may even be filled. Always assume a delay between when you add to storage and when it is actually

> recorded. <!--I'd like to understand the truck metaphor better, and maybe come up with something more accessible. What are we assuming about how a truck on a road behaves differently from fluid in a pipe?-->

Reviewer: I believe this predates any of us. @AmySteam or @oliverdunk any ideas for a better metaphor.

Reviewer 2: I think the key ideas are:

- Don't assume storage is free. There is some performance cost to storing things and so be mindful about if you need to store something or if you can store just the important things.
- There are quotas on the storage API and if you are frequently storing things/storing things that are too big you may exceed these and then lose the ability to store things entirely for a period of time.
- Storage isn't always instant, and may take some amount of time to complete.

Author: Okay, in my head all three of those points fit better with the truck metaphor than the pipe metaphor.
I think it's better if we just replace this analogy with exactly what Oliver said (but punctuated to make it slightly less casual).

Reviewer: Sounds good.

Author: Updated. Thoughts about the phrasing?
